File: cpp/day05_seed_fertilizer.cpp
// Day 5: If You Give a Seed a Fertilizer
// https://adventofcode.com/2023/day/5
#include<iostream>
#include<fstream>
#include<set>
#include<cmath>
#include<map>

using namespace std;

class Mapping {
public:
    string src;
    string dst;
    vector<pair<long, long>> src_ranges; 
    map<pair<long, long>, long> dst_ranges;
    Mapping(string src, string dst) {
        this->src = src;
        this->dst = dst;
        src_ranges = {};
        dst_ranges = {};
    }

    string get_src() {
        return this->src;
    }

    string get_dst() {
        return this->dst;
    }

    void add_range(long src_start, long dst_start, long length) {
        long src_end = src_start + length - 1;
        this->src_ranges.push_back(make_pair(src_start, src_end));
        sort(this->src_ranges.begin(), this->src_ranges.end());
        this->dst_ranges[make_pair(src_start, src_end)] = dst_start;
    }

    long get_mapping(long src) {
        pair<long, long> range = this->binary_search(src); 
        if (range.first != -1) {
            long diff = src - range.first;
            return this->dst_ranges[range] + diff;
        }
        return src;
    }

    pair<long, long> binary_search(long src) {
        int i = 0;
        int j = this->src_ranges.size();
        int m;

        while (i < j) {
            m = (i+j) / 2;
            if (this->src_ranges[m].first > src) {
                j = m;
            } else {
                i = m+1;
            }
        }

        pair<long, long> range = this->src_ranges[j-1];
        if (range.first <= src && src <= range.second) {
            return range;
        }
        return make_pair(-1, -1);
    }
};


vector<string> split_string(string s, string delimiter = " ") {
    vector<string> res;
    size_t pos = 0;
    string next;
    while ((pos = s.find(delimiter)) != string::npos) {
        next = s.substr(0, pos);
        if (next.length() != 0) {
            res.push_back(next);
        }
        s.erase(0, pos + delimiter.length());
    }
    // last item
    if (s.length() > 0) res.push_back(s);
    return res;
}


vector<long> parse_seeds(string line) {
    vector<long> res;
    vector<string> split = split_string(line);
    for (int i = 1; i < split.size(); i++) {
        res.push_back(stol(split[i]));
    }
    return res;
}


pair<string, string> parse_mapping_name(string line) {
    int i1 = line.find('-');
    int i2 = line.find('-', i1+1);
    int i3 = line.find(' ');
    return make_pair(line.substr(0, i1), line.substr(i2+1, i3-i2-1));
}

tuple<long, long, long> parse_mapping(string line) {
    vector<string> nums = split_string(line);
    return make_tuple(stol(nums[0]), stol(nums[1]), stol(nums[2]));
}



auto read_input() {
    string filename = "../inputs/day05_input.txt";
    cout << "Input file: " << filename << endl;
    string line;
    ifstream input(filename);

    // First line
    getline(input, line);
    vector<long> seeds = parse_seeds(line);

    Mapping* currmap;
    unordered_map<string, Mapping*> mappings;

    while (getline(input, line)) {
        if (line.find("map") != string::npos) {
            pair<string, string> mapnames = parse_mapping_name(line);
            currmap = new Mapping(mapnames.first, mapnames.second);
            mappings[mapnames.first] = currmap;
        } else if (line != "") {
            tuple<long, long, long> nums = parse_mapping(line);
            currmap->add_range(get<0>(nums), get<1>(nums), get<2>(nums));
        }
    }
    input.close();
    return make_pair(seeds, mappings);
}


long get_location(long seed, unordered_map<string, Mapping*>& mappings) {
    string src = "seed";
    long curr = seed;
    Mapping* currmap;
    while (src != "location") {
        cout << curr << endl;
        Mapping* currmap = mappings[src];
        curr = currmap->get_mapping(curr);
        src = currmap->get_dst();
    }
    return curr;
}


auto part1() {
    auto r = read_input();
    vector<long> seeds = r.first;
    unordered_map<string, Mapping*> mappings = r.second;

    long minloc = LONG_MAX;
    for (long seed : seeds) {
        cout << "Seed " << seed << endl;
        minloc = min(minloc, get_location(seed, mappings));
    }
    return minloc;
}


auto part2() {


    return 2;
}


int main() {
    auto p1 = part1();
    cout << "Part 1: " << p1 << endl;

    cout << "--------------" << endl;

    auto p2 = part2();
    cout << "Part 2: " << p2 << endl;
}

File: python/day05_seed_fertilizer.py
# Day 5: If You Give a Seed a Fertilizer
# https://adventofcode.com/2023/day/5

from util.input_util import read_input_file


class Mapping:
    def __init__(self, src:str, dst:str):
        self.src = src
        self.dst = dst
        self.src_ranges = [] # list of tuples
        self.dst_ranges = {} # (src_start, src_end) -> dst_start
        self.cache = {}
        
    def get_src(self) -> str:
        return self.src
    
    def get_dst(self) -> str:
        return self.dst

    def add_range(self, src_start:int, dst_start:int, length:int) -> None:
        # start, end - inclusive
        src_end = src_start+length-1
        self.src_ranges.append((src_start, src_end))
        self.src_ranges.sort()
        self.dst_ranges[(src_start, src_end)] = dst_start
    
    def get_mapping(self, src:int) -> int:
        if src in self.cache: return self.cache[src]

        start, end = self.binary_search(src)
        if start <= src <= end:
            diff = src - start
            self.cache[src] = self.dst_ranges[(start, end)] + diff
            return self.cache[src]
            
        self.cache[src] = src
        return src

        
    def binary_search(self, src:int) -> tuple:
        # find first tuple with start <= src
        # then check if in tuple
        i, j = 0, len(self.src_ranges)
        while i<j:
            m = (i+j) // 2
            # print(m, src, self.src_ranges[m])
            if self.src_ranges[m][0] > src:
                j = m
            else:
                i = m+1

        start, end = self.src_ranges[j-1]
        if start <= src <= end:
            return (start, end) 
        return (-1, -1)
    
    def __repr__(self):
        s = f"[{self.src} -> {self.dst}]\n{self.dst_ranges}"
        return s
        
        

def parse_lines():
    lines = read_input_file(5)
    seeds = [int(x) for x in lines[0].replace("seeds: ", "").split()]
    mappings = {}
    currMap = None

    for l in lines[2:]:
        if l == "":
            continue
        elif "map" in l:
            # seed-to-soil map:
            src, _, dst = l.split()[0].split('-')
            currMap = Mapping(src, dst)
            mappings[src] = currMap
        else:
            dst_start, src_start, length = l.split()
            currMap.add_range(int(src_start), int(dst_start), int(length))
            
    return seeds, mappings


# dest range start, src range start, range length
# any src numbers that aren't mapped correspond to the same dest
# 

def trace_mappings(seed:int, mappings:dict) -> int:
    src = 'seed'
    curr = seed
    while src != 'location':
        print(curr)
        m = mappings[src]
        curr = mappings[src].get_mapping(curr)
        src = m.get_dst()
    
    return curr



def solution1():
    seeds, mappings = parse_lines()
    minloc = float('inf')

    for s in seeds:
        print("Seed", s)
        loc = trace_mappings(s, mappings)
        minloc = min(minloc, loc)
    
    return minloc



    
def solution2():
    seed_ranges_raw, mappings = parse_lines()
    minloc = float('inf')

    
    # Try to get all boundary values from all mappings
    # We know min has to be on the boundary somewhere
    boundaries = [] # all distinct values to consider
    seed_ranges = [] # tuples of (start, end)
    for i in range(0, len(seed_ranges_raw), 2):
        start = seed_ranges_raw[i]
        l = seed_ranges_raw[i+1]
        end = start+l
        seed_ranges.append((start, end))
        boundaries += [start, end]
    
    # Include distinct values from the various mappings
    for _, m in mappings.items():
        for (a, b), c in m.dst_ranges.items():
            boundaries += [a, b, c]

    for s in boundaries:
        # Only consider boundary unit if within original seed ranges
        for r0, r1 in seed_ranges:
            if r0 <= s <= r1:
                loc = trace_mappings(s, mappings)
                minloc = min(minloc, loc)
    
    return minloc



if __name__ == '__main__':
    print("Part 1:", solution1())
    
    print('--------------')
    
    # print("Part 2: ", solution2())